File: src/main/java/com/gregtechceu/gtceu/api/sync_system/ContextualFieldCodec.java
package com.gregtechceu.gtceu.api.sync_system;

import net.minecraft.core.HolderLookup;
import net.minecraft.nbt.Tag;

import org.jetbrains.annotations.Nullable;

public interface ContextualFieldCodec<T> {

    Tag serializeNBT(T value, Context<T> context);

    @Nullable
    T deserializeNBT(Tag tag, Context<T> context);

    default boolean shouldSyncField(T value, Context<T> context, boolean fullSync, boolean manuallyDirty) {
        return fullSync || manuallyDirty;
    }

    record Context<T>(Object holder, TypeDeclaration type, @Nullable T currentValue, String fieldName,
                      boolean isClientSync, boolean isClientFullSyncUpdate, HolderLookup.Provider lookup) {}
}

File: src/main/java/com/gregtechceu/gtceu/api/sync_system/ContextualFieldCodec.kt
package com.gregtechceu.gtceu.api.sync_system

import net.minecraft.core.HolderLookup

import com.google.gson.JsonElement
import org.jetbrains.annotations.Nullable

enum class SyncSerializationTarget {
	NBT,
	DATA_COMPONENTS,
}

interface ContextualFieldCodec<T> {
	fun serializeField(value: T, context: Context<T>): JsonElement =
		throw UnsupportedOperationException("Sync: field ${context.fieldName} uses ${javaClass.name}, which does not support DataComponentMap serialization")

	@Nullable
	fun deserializeField(value: JsonElement, context: Context<T>): T? =
		throw UnsupportedOperationException("Sync: field ${context.fieldName} uses ${javaClass.name}, which does not support DataComponentMap deserialization")

	fun shouldSyncField(value: T, context: Context<T>, fullSync: Boolean, manuallyDirty: Boolean): Boolean = fullSync || manuallyDirty

	@JvmRecord
	data class Context<T> @JvmOverloads constructor(
		val holder: Any,
		val type: TypeDeclaration,
		@field:Nullable val currentValue: T?,
		val fieldName: String,
		val isClientSync: Boolean,
		val isClientFullSyncUpdate: Boolean,
		val lookup: HolderLookup.Provider,
		val serializationTarget: SyncSerializationTarget = SyncSerializationTarget.DATA_COMPONENTS,
		val parseExplicitNull: Boolean = false,
	)
}
